django/db/models/fields/related.py

```diff
@@ -854,6 +856,59 @@ def _check_on_delete(self):
                     id='fields.E321',
                 )
             ]
+        elif on_delete == DB_CASCADE and (
+            any(  # generic relation
+                hasattr(field, 'bulk_related_objects') for field in
+                self.model._meta.private_fields
+            ) or
+            any(  # generic foreign key
+                hasattr(field, 'get_content_type') for field in
+                self.model._meta.private_fields
+            )
+        ):
+            return [
+                checks.Warning(
+                    'Field specifies unsupported DB_CASCADE on model '
```

Comment:
I think all of these checks need to be revisited to check whether we also need to support warning about `DB_RESTRICT` and `DB_SET_NULL`. There is no need to define extra id values, e.g. `W346` could be reused to support a warning for both `DB_CASCADE` and `DB_RESTRICT` for example - just make sure the related documentation reflects this.